test/test.js

```diff
@@ -0,0 +1,45 @@
+const nock = require('nock')
+const got = require('got')
+const App = require('/path/to/app')
+
+describe('app.js', function() {
+  let app
+
+  beforeEach(function() {
+    // set up stuff
+    app = new App({
+      id: 123, privateKey: 'secrets'
+    })
+  });
+  // test cases
+
+  // Get app installations
+  // https://developer.github.com/v3/apps/#find-repository-installation
+  it('gets an individual installation', () => {
+    nock('https://apps-js-test.com', {
+      reqheaders: {
+        authorization: `token ${app.appAuth()}`, // jwt: installation ID to auth as an installation
```

Comment:
mocks should always be the same. And I think authenticating as app requires the prefix to be `Bearer ` not `token `, see https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#authenticating-as-a-github-app

I’d do the following:

For one of your apps, generate a new private key. Store it in the test folder. You can commit it later, but make sure to remove the key from your app before that.

Using that key send, send a real, valid request to list the app’s installations. That way you can record both the request and the response

The code will look something like this:

```js
const {readFileSync} = require('fs')
const jwt = require('jsonwebtoken')

const cert = fs.readFileSync('private.key')
const payload = {
  exp: Math.floor(Date.now() / 1000) + 60,  // JWT expiration time
  iat: Math.floor(Date.now() / 1000),       // Issued at time
  iss: 123                                  // Your GitHub App’s ID
}
const token = jwt.sign(payload, cert, {algorithm: 'RS256'})

got('https://api.github.com/repos/hiimbex/testing-things/installation', {
  headers: {
    accept: 'application/vnd.github.machine-man-preview+json',
    authorization: `Bearer ${token}`
  }
}).then(response => console.log(response.body))
```

Once you have that working, record the request/response with nock by adding this to the top

```js
const nock = require('nock')
nock.recorder.rec()
```

When you run your script again, you should now see fixtures printed to the terminal that will be ready to copy & paste. Use these to mock the request using `nock('https://api.github.com').get(...` etc. Remove clutter that you don’t need from the fixtures, only leave in what’s important. Comment out the `nock.recorder.rec()` statement, see if your script still works.

When it does, great, make sure to commit the state so you don’t loose it. Next I would remove the dynamic setting of `payload.exp` and `payload.iat` using `Date.now()`, instead set it to `0` and `60`. Log out the resulting `token`, put that token into the fixutres `reqheaders.authorization`. 

Once that’s all working, build the actual test with something like mocha and instead of using `got`, use code as it’s shown in the readme. You will need to mock time, so that `Date.now()` in the code will always return `0`. You can do that with https://www.npmjs.com/package/lolex. Make the code work until the test passes.

Rinse & repeat for the "create an issue" test :)

Reply:
Just documenting my progress here:

I got it working for real and have correct logs and api response for that part, but whenever I add in the `nock.recorder.rec()` everything runs, but there's nothing in the console, and I can't seem to debug it, since [nock's default](https://github.com/nock/nock#recording) is to log all responses to the console.

So I'm a bit stuck here.

Reply:
could you share the code for the script you use the record fixtures?

Reply:
```js
const nock = require('nock')
const got = require('got')
//const App = require('/path/to/app')
const fs = require('file-system')
const jwt = require('jsonwebtoken')

//nock.recorder.rec()

const cert = fs.readFileSync(`${__dirname}/pem/private-key.pem`)

const payload = {
  exp: Math.floor(Date.now() / 1000) + 60,  // JWT expiration time
  iat: Math.floor(Date.now() / 1000),       // Issued at time
  iss: 3698                                  // Your GitHub App’s ID
}
const token = jwt.sign(payload, cert, {algorithm: 'RS256'})

got('https://api.github.com/repos/hiimbex/testing-things/installation', {
  headers: {
    accept: 'application/vnd.github.machine-man-preview+json',
    authorization: `Bearer ${token}`
  }
}).then(response => console.log(response.body))
.catch(err => console.log(err))
```

Correctly outputs:
```json
{"id":39001,"account":{"login":"hiimbex","id":13410355,"node_id":"MDQ6VXNlcjEzNDEwMzU1","avatar_url":"https://avatars1.githubusercontent.com/u/13410355?v=4","gravatar_id":"","url":"https://api.github.com/users/hiimbex","html_url":"https://github.com/hiimbex","followers_url":"https://api.github.com/users/hiimbex/followers","following_url":"https://api.github.com/users/hiimbex/following{/other_user}","gists_url":"https://api.github.com/users/hiimbex/gists{/gist_id}","starred_url":"https://api.github.com/users/hiimbex/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/hiimbex/subscriptions","organizations_url":"https://api.github.com/users/hiimbex/orgs","repos_url":"https://api.github.com/users/hiimbex/repos","events_url":"https://api.github.com/users/hiimbex/events{/privacy}","received_events_url":"https://api.github.com/users/hiimbex/received_events","type":"User","site_admin":false},"repository_selection":"selected","access_tokens_url":"https://api.github.com/installations/39001/access_tokens","repositories_url":"https://api.github.com/installation/repositories","html_url":"https://github.com/settings/installations/39001","app_id":3698,"target_id":13410355,"target_type":"User","permissions":{"pull_requests":"write","issues":"write","metadata":"read"},"events":["pull_request"],"created_at":"2017-07-12T20:26:00.000Z","updated_at":"2018-07-06T18:19:01.000Z","single_file_name":null}
```

But when `nock.recorder.rec()` is uncommented, there are no logs.